### PR TITLE
Implement Basic Dutch Auction

### PR DESCRIPTION
**Pull request by dekz**
## Description

[Quip doc here](https://0xproject.quip.com/omWdAsRTDxjh/Dutch-Auctions-on-0x-v2) for further details.

image

A simple `matchOrders` in a extension contract that validates the correct amount given the current block timestamp. `sellOrder` is the smallest possible amount for the asset up for sale. That is the `sellOrder.takerAssetAmount` is the final amount at the end of the auction. The contract guarantees the incoming buyOrder is for an greater than or equal to the calculated `currentAmount`. The `currentAmount` is calculated from the `auctionBeginTimeSeconds`, `auctionBeginAmount`, `sellOrder.takerAssetAmount` aka `auctionEndAmount` and `sellOrder.expiryTimeSeconds` aka `auctionEndTimeSeconds`. currentAmount travels linearly downwards from the `auctionBeginAmount` to the `auctionEndAmount`. 

Match Orders guarantees the seller gets the minimum reserve amount (`sellOrder.takerAssetAmount`). When matched the delta between `currentAmount` and `sellOrder.takerAssetAmount` is transferred to the seller. Any further excess is returned to the buyer (the delta between `buyOrder.makerAssetAmount` and `currentAmount`. This means the buyer can send in any amount safely and be assured they will get the exact price given the current block timestamp.

## Checklist:

*   [ ] Prefix PR title with `[WIP]` if necessary.
*   [ ] Prefix PR title with bracketed package name(s) corresponding to the changed package(s). For example: `[sol-cov] Fixed bug`.
*   [ ] Add tests to cover changes as needed.
*   [ ] Update documentation as needed.
*   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        );
+        // Return any spread to the seller
+        uint256 leftMakerAssetSpreadAmount = matchedFillResults.leftMakerAssetSpreadAmount;
+        if (leftMakerAssetSpreadAmount > 0) {
+            // Calculate the excess from the buy order. This can occur if the buyer sends in a higher
+            // amount than the calculated current amount
+            uint256 buyerExcessAmount = buyOrder.makerAssetAmount-auctionDetails.currentAmount;
+            uint256 sellerExcessAmount = leftMakerAssetSpreadAmount-buyerExcessAmount;
+            bytes memory assetData = sellOrder.takerAssetData;
+            address token = assetData.readAddress(16);
```
Comment (hysz) [inline packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol]: Nit - include an offset table for reference. [Example](https://github.com/0xProject/0x-monorepo/blob/development/packages/contracts/contracts/protocol/AssetProxy/ERC20Proxy.sol#L67).

File: packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            buyOrder,
+            sellOrder,
+            buySignature,
+            sellSignature
+        );
+        // Return any spread to the seller
+        uint256 leftMakerAssetSpreadAmount = matchedFillResults.leftMakerAssetSpreadAmount;
+        if (leftMakerAssetSpreadAmount > 0) {
+            // Calculate the excess from the buy order. This can occur if the buyer sends in a higher
+            // amount than the calculated current amount
```
Comment (hysz) [inline packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol]: Should probably use `safeMath` here or include a comment about why it's definitely not needed.

File: packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    /// @param order The sell order
+    /// @return AuctionDetails
+    function getAuctionDetails(
+        LibOrder.Order memory order
+    )
+        public
+        returns (AuctionDetails memory auctionDetails)
+    {
+        uint256 makerAssetDataLength = order.makerAssetData.length;
+        // We assume auctionBeginTimeSeconds and auctionBeginAmount are appended to the makerAssetData
```
Comment (hysz) [inline packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol]: Same note as above - should include an offset table for reference. [Example](https://github.com/0xProject/0x-monorepo/blob/development/packages/contracts/contracts/protocol/AssetProxy/ERC20Proxy.sol#L67).
Could probably just include this once in here then reference it above.

File: packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        uint256 auctionBeginTimeSeconds = order.makerAssetData.readUint256(makerAssetDataLength-64);
+        uint256 auctionBeginAmount = order.makerAssetData.readUint256(makerAssetDataLength-32);
+        require(order.expirationTimeSeconds > auctionBeginTimeSeconds, "INVALID_BEGIN_TIME");
```
Comment (hysz) [inline packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol]: Same note about `safeMath` in this function.

File: packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        uint256 currentTimeSeconds;  // block.timestamp
+    }
+
+    constructor (address _exchange)
+        public
+    {
+        EXCHANGE = IExchange(_exchange);
+    }
+
+    /// @dev Matches the buy and sell orders at an amount given the following: the current block time, the auction
```
Comment (hysz) [inline packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol]: Nice

File: packages/contracts/test/extensions/dutch_auction.ts [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+                    auctionBeginTimeSeconds,
+                    auctionBeginAmount,
+                ),
+            });
+            const auctionDetails = await dutchAuctionContract.getAuctionDetails.callAsync(sellOrder);
+            expect(auctionDetails.currentAmount).to.be.bignumber.equal(auctionBeginAmount);
+            expect(auctionDetails.beginAmount).to.be.bignumber.equal(auctionBeginAmount);
+        });
+        it('should be be worth the end price at the end of the auction', async () => {
+            auctionBeginTimeSeconds = new BigNumber(currentBlockTimestamp - 1000);
```
[the remark this answers is not in the document]
Comment (dekz) [reply packages/contracts/test/extensions/dutch_auction.ts]: I'll re-use tenMinutesInSeconds

File: packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        // solhint-disable-next-line not-rely-on-time
+        uint256 timestamp = block.timestamp;
+        auctionDetails.beginTimeSeconds = auctionBeginTimeSeconds;
+        auctionDetails.endTimeSeconds = order.expirationTimeSeconds;
+        auctionDetails.beginAmount = auctionBeginAmount;
+        auctionDetails.endAmount = minAmount;
+        auctionDetails.currentTimeSeconds = timestamp;
+
+        uint256 remainingDurationSeconds = order.expirationTimeSeconds-timestamp;
+        uint256 currentAmount = minAmount + (remainingDurationSeconds*amountDelta/auctionDurationSeconds);
```
Comment (abandeali1) [inline packages/contracts/contracts/extensions/DutchAuction/DutchAuction.sol]: Nit: add spaces in between arithmetic operators

**Pull request merged**